Enable building with CUDA support on Mac OS X

ville-k opened the pull request:
- building with CUDA support on OS X requires GNU coreutils due to the OS X
  native readlink command behaving differently from the GNU version - you can
  install it using homebrew: "brew install coreutils"
- OS X CUDA builds against CUDA toolkit v. 7.5 to overcome host compiler
  incompatibility - the toolkit versions (CUDA & cuDNN) are now controlled by
  variables set in the configure script
- cuda/platform.bzl file is dynamically generated by the configure script to
  overcome bazel limitations for using "select" to set the platform
  specific names and paths of CUDA libraries
- SE_STATIC_THREAD_LOCAL_POD now uses __thread instead of thread_local which
  is not yet supported by the version of clang shipped by Apple. __thread
  supports only primitive types, but is more performant
- Updates Eigen to a newer version that fixes a (clang) compilation error

ville-k committed:
Enable building with CUDA support on Mac OS X

* Building with CUDA support on OS X requires GNU coreutils due to the OS X
  native readlink command behaving differently from the GNU version
* OS X requires CUDA toolkit 7.5 due to host compiler incompatibility
  with earlier CUDA toolkits - the toolkit versions (CUDA & cuDNN) are now
  controlled by variables set in the configure script

Mistobaan commented:
Nice! I think this is clearly and improvement over the existing one. I tried on my macbook with cuda and it works. I also published a [small tutorial](https://medium.com/@fabmilo/how-to-compile-tensorflow-with-cuda-support-on-osx-fd27108e27e1#.kaq068rjh) on how to use this patch]. Hope it can be merged soon. @vrv @mrry thoughts ?

ville-k commented:
Thanks for trying this out @Mistobaan and writing a tutorial!  I'd add a step for installing GNU coreutils ("brew install coreutils") to the tutorial - most people probably don't have it installed.

vrv commented:
This is very nice!  Thanks for this contribution -- we'll have @zheng-xq take a look at this soon.

Mistobaan commented:
@ville-k good point. Updated :)

zheng-xq commented:
@leary-google, could you review the stream-execuctor portion of this change?

fpmchu commented:
@Mistobaan I followed your instructions, but "brew cask install cuda" defaults to CUDA 7.0, and @ville-k patch is using 7.5 by default.  I tried both.  With CUDA 7.0, I get a compile error as such:
```
INFO: From Compiling tensorflow/core/kernels/cwise_op_gpu_sin.cu.cc:
nvcc fatal   : The version ('70002') of the host compiler ('Apple clang') is not supported
ERROR: /Users/fpmc/git/tensorflow/tensorflow/core/BUILD:339:1: error while parsing .d file: /private/var/tmp/_bazel_fpmc/b41e6b4c9df9b99106d3673ec4f590dc/tensorflow/bazel-out/local_darwin-opt/bin/tensorflow/core/_objs/gpu_kernels/tensorflow/core/kernels/cwise_op_gpu_sin.cu.d (No such file or directory).
```

By manually downloading CUDA 7.5 and installing it, it compiles.

fpmchu commented:
@ville-k In testing your change I find that the newly added ALT_PATH doesn't really work.  If you have ```libcudnn.6.5.dylib``` located inside ```/usr/local/cuda/``` and not ```/usr/local/cuda/lib/```, the symlink command at the end of your ```cuda_config.sh``` will end up silently creating a bad symlink like this

```
libcudnn.6.5.dylib@ -> /usr/local/cuda/lib/libcudnn.6.5.dylib/lib/libcudnn.6.5.dylib
```

I think having the ALT_PATH stuff is hard to get right.

ville-k committed:
Fix symlink creation for cuDNN when found in alternative location.

ville-k commented:
@fpmchu Thanks for testing the ALT_PATH build scenario for cuDNN!  The fix for the problem you discovered turned out to be pretty simple - I have it on a separate branch for now: 
https://github.com/ville-k/tensorflow/commit/1232b37d3596a154a74e162db33c888111eba17d

@vrv What's the project's policy for issues found and fixed during PR review?  New commits on the existing PR or open a separate PR?

vrv commented:
New commits on existing PR seems fine -- we'll just ask to squash the commits prior to validation and merging.

fpmchu commented:
@ville-k Cool.  While the fix looks ok, I'm still not understanding the purpose of adding ALT_PATH.  Is it just to allow users to put them in /usr/local/lib?  What's wrong with using /usr/local/cuda/lib?

ville-k commented:
@fpmchu ALT_PATH is there to support existing lib search path functionality for both linux and mac. The original configure and cuda_config.sh scripts look for cudnn.so.6.5 under both "/usr/local/cuda" and "/usr/local/cuda/lib64". Depending on the platform, these locations will now be searched if the user inputs "/usr/local/cuda" as the cudnn install dir:
**Linux**
* /usr/local/cuda/lib64/cudnn.so.6.5 
* /usr/local/cuda/cudnn.so.6.5  (ALT_PATH)

**Mac**
* /usr/local/cuda/lib/cudnn.6.5.dylib
* /usr/local/cuda/cudnn.6.5.dylib  (ALT_PATH)

Mistobaan commented:
Thanks @fpmchu to try the tutorial out. I think it installed 7.0 because you have an old cuda formula. I will update the tutorial to suggest to update homebrew first and check for the cuda version.

fpmchu commented:
Thanks @Mistobaan.  I actually think you mean ```brew update``` though.  I did try "upgrade" before and that didn't work.  I didn't know that "update" is the thing to do to "upgrade brew" :-)

martinwicke commented:
Because of peculiarities in our internal build process, we won't be able to merge this right now. I'll leave this open since it may be useful for people. When we find someone to resolve the internal problems, we may be able to absorb it at a later time.

I'm sorry about that -- I would love to have this in.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

ville-k commented:
Thanks for the update @martinwicke ! Is the main issue causing problems with your internal build process the automatic generation of the "platform.bzl" file?

martinwicke commented:
That, and reuse of shared code elsewhere. The code elsewhere can be updated
(internally) which is why we need someone here to fix that up. The problem
with stream_executor is that it almost has to be treated like generated
code. In sorry this bit you after you've done all this work, I'm still
hoping we can absorb it somehow, and we'll make the restrictions on
stream_executor clearer for the future.
On Fri, Jan 8, 2016 at 07:25 Ville Kallioniemi <notifications@github.com>
wrote:

> Thanks for the update @martinwicke <https://github.com/martinwicke> ! Is
> the main issue causing problems with your internal build process the
> automatic generation of the "platform.bzl" file?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/664#issuecomment-170030821>
> .
>

NathanHowell commented:
I get a segfault when the  cuda libs are not setup properly due to `getenv("LD_LIBRARY_PATH")` returning null:

```
(lldb) bt
* thread #1: tid = 0x30ae49, 0x00007fff8ed8e752 libsystem_c.dylib`strlen + 18, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
  * frame #0: 0x00007fff8ed8e752 libsystem_c.dylib`strlen + 18
    frame #1: 0x0000000105f0d7a4 _pywrap_tensorflow.so`std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::operator<<<std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, char const*) [inlined] std::__1::char_traits<char>::length(__s=0x0000000000000000) + 52 at string:651
    frame #2: 0x0000000105f0d790 _pywrap_tensorflow.so`std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::operator<<<std::__1::char_traits<char> >(__os=0x00007fff5fbfbb10, __str=0x0000000000000000) + 32 at ostream:882
    frame #3: 0x000000010895c96a _pywrap_tensorflow.so`perftools::gputools::internal::DsoLoader::GetDsoHandle(tensorflow::StringPiece, void**, perftools::gputools::internal::DsoLoader::LoadKind) [inlined] std::__1::enable_if<(__os=0x00007fff5fbfbb10, __x=0x00007fff5fbfbb08)) && (is_base_of<std::__1::ios_base, tensorflow::internal::LogMessage>::value), tensorflow::internal::LogMessage&&>::type std::__1::operator<<<tensorflow::internal::LogMessage, char*>(tensorflow::internal::LogMessage&&, char* const&) + 19 at ostream:1057
<trimmed>
(lldb) frame select 4
frame #4: 0x000000010895c957 _pywrap_tensorflow.so`perftools::gputools::internal::DsoLoader::GetDsoHandle(path=(data_ = "libcuda.dylib", size_ = 13), dso_handle=0x00007fff5fbfc138, load_kind=kLocal) + 887 at dso_loader.cc:99
   96  	  string path_string = path.ToString();
   97  	  *dso_handle = dlopen(path_string.c_str(), dynload_flags);
   98  	  if (*dso_handle == nullptr) {
-> 99  	    LOG(INFO) << "Couldn't open CUDA library " << path
   100 	              << ". LD_LIBRARY_PATH: " << getenv("LD_LIBRARY_PATH");
   101 	    // TODO(b/22689637): Eliminate unnecessary ToString once StrCat has been
   102 	    // moved to the open-sourceable version.
```

NathanHowell commented:
I eventually did get this working but the version of Eigen referenced in here is very broken. Eigen HEAD (fd9611fa2d9c) does work aside from a nvcc build break in TensorIntDiv.h, `DividerHelper<64, T>::computeMultiplier` is missing a cast... but it does at least seem to work.

Previous failure looked like this:

> libc++abi.dylib: terminating with uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument

After a bit of hunting around it turns out that the mutex instances had already been destructed after someone called `exit(1)` :hushed: 

```
(lldb) bt
* thread #2: tid = 0x4649e, 0x00007fff98e3f738 libsystem_c.dylib`exit, stop reason = breakpoint 1.1
  * frame #0: 0x00007fff98e3f738 libsystem_c.dylib`exit
    frame #1: 0x00000001066e9f0e _pywrap_tensorflow.so`void Eigen::internal::EigenMetaKernel_Vectorizable<Eigen::TensorEvaluator<Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<float, 1, 1, int>, 16>, Eigen::TensorCwiseUnaryOp<Eigen::internal::scalar_right<float, float, Eigen::internal::scalar_difference_op<float>, true>, Eigen::TensorMap<Eigen::Tensor<float const, 1, 1, int>, 16> const> const> const, Eigen::GpuDevice>, int>(Eigen::TensorEvaluator<Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<float, 1, 1, int>, 16>, Eigen::TensorCwiseUnaryOp<Eigen::internal::scalar_right<float, float, Eigen::internal::scalar_difference_op<float>, true>, Eigen::TensorMap<Eigen::Tensor<float const, 1, 1, int>, 16> const> const> const, Eigen::GpuDevice>, int) + 14
    frame #2: 0x00000001066e6ec1 _pywrap_tensorflow.so`tensorflow::functor::BinaryFunctor<Eigen::GpuDevice, tensorflow::functor::sub<float>, 1>::Right(Eigen::GpuDevice const&, Eigen::TensorMap<Eigen::Tensor<float, 1, 1, long>, 16>, Eigen::TensorMap<Eigen::Tensor<float const, 1, 1, long>, 16>, Eigen::TensorMap<Eigen::TensorFixedSize<float const, Eigen::Sizes<>, 1, long>, 16>) + 257
```

And it turns out that all the `EigenMetaKernel_Vectorizable` specializations don't work as intended:

```
(lldb) disassemble
_pywrap_tensorflow.so`void Eigen::internal::EigenMetaKernel_Vectorizable<Eigen::TensorEvaluator<Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<float, 1, 1, int>, 16>, Eigen::TensorCwiseUnaryOp<Eigen::internal::scalar_right<float, float, Eigen::internal::scalar_difference_op<float>, true>, Eigen::TensorMap<Eigen::Tensor<float const, 1, 1, int>, 16> const> const> const, Eigen::GpuDevice>, int>:
    0x1066e9f00 <+0>:  pushq  %rbp
    0x1066e9f01 <+1>:  movq   %rsp, %rbp
    0x1066e9f04 <+4>:  movl   $0x1, %edi
    0x1066e9f09 <+9>:  callq  0x1069fe13e               ; symbol stub for: exit
    0x1066e9f0e <+14>: nop
```

ville-k commented:
I'm not able to reproduce the issue with LD_LIBRARY_PATH you mentioned @NathanHowell Which version of OSX and Python are you using?
I reported the missing cast issue to Eigen before this PR and they fixed it for the version that I'm using (their fix was to add a constructor, not cast the argument). Sounds like they might've had some regression if it's broken in their HEAD. I'll push an update to this PR today with latest from master and try to see if I can find a rev of Eigen that is never and does not have have regression.

NathanHowell commented:
@ville-k the segfault might have been from an older version of Xcode, I upgraded to 7.2 trying to track down the other issue and I think it's been fixed... but it should use `DYLD_LIBRARY_PATH` on osx rather than `LD_LIBRARY_PATH` right?

ville-k commented:
@NathanHowell I was surprised about LD_LIBRARY_PATH working too when I stumbled into it working by accident. Apple added this at some point for UNIX conformance - they're checking both env vars nowadays:
http://www.opensource.apple.com/source/dyld/dyld-360.18/src/dyld.cpp

elbamos commented:
Worked for me as well with @Mistobaan 's instructions.  Note that `coreutils` has to be installed from brew even if you think it's already installed, because the installer calls `greadlink`, so the brew package is a hard dependency on OS X.  In addition, there should be a better way of handling the `LD_LIBRARY_PATH` issue.  Setting any of those environment variables manually can wreak all sorts of havok on OS X.  Neither theano nor torch needs it set explicitly.

elbamos commented:
@ville-k would you mind terribly rebasing?  On linux, updates to bazel have created a number of installation issues where fixes were rolled into the git in the last few weeks.

NathanHowell commented:
@elbamos I have a rebased/hacked up branch here: https://github.com/NathanHowell/tensorflow/tree/cuda_osx3

It also needs a patch to Eigen, not sure if it's 100% correct but it at least builds and runs:

```
--- ./bazel-tensorflow/external/eigen_archive/eigen-eigen-c8e5d094f3a9/unsupported/Eigen/CXX11/src/Tensor/TensorIntDiv.h	2016-01-25 15:51:44.000000000 -0800
+++ ./bazel-tensorflow/external/eigen_archive/eigen-eigen-c8e5d094f3a9/unsupported/Eigen/CXX11/src/Tensor/TensorIntDiv.h	2016-01-25 16:04:11.000000000 -0800
@@ -98,7 +98,7 @@
       return static_cast<uint64_t>((static_cast<__uint128_t>(1) << (64+log_div)) / static_cast<__uint128_t>(divider) - (static_cast<__uint128_t>(1) << 64) + 1);
 #else
       const uint64_t shift = 1ULL << log_div;
-      TensorUInt128<uint64_t, uint64_t> result = (TensorUInt128<uint64_t, static_val<0> >(shift, 0) / TensorUInt128<static_val<0>, uint64_t>(divider) - TensorUInt128<static_val<1>, static_val<0> >(1, 0) + TensorUInt128<static_val<0>, static_val<1> >(1));
+      TensorUInt128<uint64_t, uint64_t> result = (TensorUInt128<uint64_t, static_val<0> >(shift, 0) / TensorUInt128<static_val<0>, uint64_t>(static_cast<uint64_t>(divider)) - TensorUInt128<static_val<1>, static_val<0> >(1, 0) + TensorUInt128<static_val<0>, static_val<1> >(1));
       return static_cast<uint64_t>(result);
 #endif
     }
```

ville-k commented:
@elbamos I started a rebase over the weekend, but ran into the Eigen issue also reported here: https://github.com/tensorflow/tensorflow/issues/883
I'll give it another try tomorrow.

ville-k committed:
* fix library name formatting broken during rebase
* point to Eigen fork that builds on OS X - pending Eigen PR merge

ville-k commented:
@elbamos I just rebased and pointed this PR to temporarily use my fork of Eigen that builds on OSX (Eigen PR is pending)

elbamos commented:
@ville-k but I thought it was working on OS X as it was?

elbamos commented:
Anyway here's what I get now:

```
INFO: From Compiling tensorflow/core/kernels/xent_op_gpu.cu.cc:
nvcc warning : option '--relaxed-constexpr' has been deprecated and replaced by option '--expt-relaxed-constexpr'.
nvcc warning : option '--relaxed-constexpr' has been deprecated and replaced by option '--expt-relaxed-constexpr'.
/usr/lib/gcc/x86_64-linux-gnu/5/include/mwaitxintrin.h(36): error: identifier "__builtin_ia32_monitorx" is undefined

/usr/lib/gcc/x86_64-linux-gnu/5/include/mwaitxintrin.h(42): error: identifier "__builtin_ia32_mwaitx" is undefined

2 errors detected in the compilation of "/tmp/tmpxft_000041a1_00000000-10_xent_op_gpu.cu.compute_52.cpp1.ii".
ERROR: /mnt/hfsshare/LIBRARIES/tensorflow/tensorflow/core/BUILD:339:1: output 'tensorflow/core/_objs/gpu_kernels/tensorflow/core/kernels/xent_op_gpu.cu.o' was not created.
ERROR: /mnt/hfsshare/LIBRARIES/tensorflow/tensorflow/core/BUILD:339:1: not all outputs were created.
Target //tensorflow/cc:tutorials_example_trainer failed to build
INFO: Elapsed time: 34.791s, Critical Path: 34.04s
```

ville-k committed:
Do not auto generate platform.bzl - modify it as part of unofficial settings.

Geka000 commented:
I have this error on mac os x

    tensorflow/stream_executor/cuda/cuda_dnn.cc:843:48: error: expected body of lambda expression
      auto get_algorithm = [&](bool specify_limit) SHARED_LOCKS_REQUIRED(
                                                   ^
    ./tensorflow/core/platform/default/thread_annotations.h:83:3: note: expanded from macro 'SHARED_LOCKS_REQUIRED'
      THREAD_ANNOTATION_ATTRIBUTE__(shared_locks_required(__VA_ARGS__))
      ^
    ./tensorflow/core/platform/default/thread_annotations.h:42:42: note: expanded from macro 'THREAD_ANNOTATION_ATTRIBUTE__'
    #define THREAD_ANNOTATION_ATTRIBUTE__(x) __attribute__((x))

NathanHowell commented:
This is due to an old version of Clang/XCode. You probably just need to
upgrade to XCode 7.x.
On Feb 3, 2016 07:12, "Geka000" <notifications@github.com> wrote:

> I have this error on mac os x
>
> tensorflow/stream_executor/cuda/cuda_dnn.cc:843:48: error: expected body of lambda expression
>   auto get_algorithm = [&](bool specify_limit) SHARED_LOCKS_REQUIRED(
>                                                ^
> ./tensorflow/core/platform/default/thread_annotations.h:83:3: note: expanded from macro 'SHARED_LOCKS_REQUIRED'
>   THREAD_ANNOTATION_ATTRIBUTE__(shared_locks_required(__VA_ARGS__))
>   ^
> ./tensorflow/core/platform/default/thread_annotations.h:42:42: note: expanded from macro 'THREAD_ANNOTATION_ATTRIBUTE__'
> #define THREAD_ANNOTATION_ATTRIBUTE__(x) __attribute__((x))
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/664#issuecomment-179285763>
> .
>

ville-k committed:
Upgrade to latest Eigen and fix issues introduced during rebase.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

ville-k commented:
@martinwicke I changed this PR to use a checked in platform.bzl file rather than autogenerating it in the configure script. Hopefully this will no longer break your internal build system and will allow this PR to be merged.  I also [worked with](https://bitbucket.org/eigen/eigen/pull-requests/158/add-constructor-for-long-types/diff) the Eigen project to fix a compilation error that was preventing Eigen from building under Cuda 7.5 - this PR branch points to the latest Eigen with that fix. Here's a quick summary of how building under OSX with/without Cuda support now works:

* when TF_UNOFFICIAL_SETTING=1 env variable is set while running the configure script, the configure script replaces the platform name, CUDA version and cuDNN constants inside platform.bzl to match user selections.  For OSX, you need to select Cuda version 7.5.

* unmodified platform.bzl file is used when configure is run in regular mode (i.e. not using unofficial settings) - OSX Cuda support will not work in this case.

ville-k commented:
Hey @martinwicke let me know if there is there are any more changes needed to merge this PR. I'll be happy to address any concerns.

ageitgey commented:
I built the latest version of this PR successfully. Thanks @ville-k!

Using @Mistobaan's gist as a starting point, I wrote updated/expanded build instructions for those less familiar with building tensorflow:

https://gist.github.com/ageitgey/819a51afa4613649bd18

Hope that helps out someone who is looking to try this.

ville-k commented:
Hey @vrv or @martinwicke could you guys give some feedback based on the changes I made and summarized in my previous comment.

vrv commented:
I think the biggest problem mentioned in https://github.com/tensorflow/tensorflow/pull/664#issuecomment-170044135 is the stream executor bits -- we don't own that code and it gets upstreamed a lot, so even if we accepted the change, it would likely get overwritten by our next upstreaming :(

We'll try to figure out a better story for stream_executor soon -- the same problem exists for Windows too.

normanhh3 commented:
@ageitgey your [instructions] (https://gist.github.com/ageitgey/819a51afa4613649bd18) worked great! I'm now using my GeForce GT 750M with 2GB of RAM to train with. :-) (Did this on OS X Yosemite - 10.10.5 - MacBook Pro Retina 15-inch Mid 2014)

matspetter commented:
I have also followed the instruction above and all the building activities works fine but when I start python and try to import tensorflow I get this error:
```
MatsMacBookPro2:~ mats$ python
Python 2.7.10 (default, Oct 23 2015, 18:05:06) 
[GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.59.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Python/2.7/site-packages/tensorflow/__init__.py", line 23, in <module>
    from tensorflow.python import *
  File "/Library/Python/2.7/site-packages/tensorflow/python/__init__.py", line 37, in <module>
    from tensorflow.core.framework.graph_pb2 import *
  File "/Library/Python/2.7/site-packages/tensorflow/core/framework/graph_pb2.py", line 10, in <module>
    from google.protobuf import descriptor_pb2
  File "/Library/Python/2.7/site-packages/google/protobuf/descriptor_pb2.py", line 1533, in <module>
    __module__ = 'google.protobuf.descriptor_pb2'
  File "/Library/Python/2.7/site-packages/google/protobuf/reflection.py", line 123, in __new__
    new_class = superclass.__new__(cls, name, bases, dictionary)
TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
```

I have a MacBook Pro Retina, 15-inch, Late 2013, GeForce GT 750M, 16GB RAM, running OS X 10.11.3

martinwicke commented:
Check the installation instructions on the website for this one -- it's a
common issue, probably reinstalling protobuf will fix it.

On Thu, Feb 25, 2016 at 1:52 PM Mats Berggrund <notifications@github.com>
wrote:

> I have also followed the instruction above and all the building activities
> works fine but when I start python and try to import tensorflow I get this
> error:
>
> MatsMacBookPro2:~ mats$ python
> Python 2.7.10 (default, Oct 23 2015, 18:05:06)
> [GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.59.5)] on darwin
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import tensorflow
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/Library/Python/2.7/site-packages/tensorflow/__init__.py", line 23, in <module>
>     from tensorflow.python import *
>   File "/Library/Python/2.7/site-packages/tensorflow/python/__init__.py", line 37, in <module>
>     from tensorflow.core.framework.graph_pb2 import *
>   File "/Library/Python/2.7/site-packages/tensorflow/core/framework/graph_pb2.py", line 10, in <module>
>     from google.protobuf import descriptor_pb2
>   File "/Library/Python/2.7/site-packages/google/protobuf/descriptor_pb2.py", line 1533, in <module>
>     __module__ = 'google.protobuf.descriptor_pb2'
>   File "/Library/Python/2.7/site-packages/google/protobuf/reflection.py", line 123, in __new__
>     new_class = superclass.__new__(cls, name, bases, dictionary)
> TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
>
> I have a MacBook Pro Retina, 15-inch, Late 2013, GeForce GT 750M, 16GB
> RAM, running OS X 10.11.3
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/664#issuecomment-189002621>
> .
>

matspetter commented:
unfortunately I have tried all that. From what I can see I have the correct protobuf installed:
>>> import google.protobuf
>>> print google.protobuf.__version__
3.0.0a3

martinwicke commented:
@matspetter Can you try again after updating protobuf again (to b2)? This may be resolved now.

esd100 commented:
Well, I spent about 10 hours trying to get this to work before finding this thread. I had to make a 50 changes and realized that you had actually made virtually all the same ones, plus another 50!

I followed your instructions (except brew cask install cuda because I had already installed that earlier and tested it with the CUDA toolkit and it was working). Unfortunately, even though I got way further than I ever had before, it failed again, with the first command bazel build -c opt --config=cuda //tensorflow/cc:tutorials_example_trainer.

It failed with 15 errors generated, almost all of them in tensorflow/stream_executor/cuda/cuda_dnn.cc

This is so frustrating. I don't understand why it's so hard to get this to work for macs.

esd100 commented:
Is there any progress on checks for this? I notice that it has been over 90 days since the original post.

Also, is there any reason that bazel should not be compiling with cuda_dnn v5?

martinwicke commented:
For cudnn v5 see #1786.

matspetter commented:
actually what I did was to make sure which versions of python I had installed. I had a whole mess of different versions from macports, osx itself (?), homebrew etc. I removed it all and reinstalled all from homebrew. Then I got it working

omarev-git commented:
I have a MBP 2015, a Vidock + with external Nvidia 680 and running El Capitan (10.11.3), and the branch work fine. Really good contribution, thank you ville-k !!!

vrv commented:
@ville-k: any interest in reviving this CL against master?  I think we'd like to merge these changes and then we'll try to keep it working as best we can.

ville-k commented:
@vrv Absolutely!  I'll start rebasing tonight.

omarev-git commented:
Fantastic!!!

Sent from my iPhone

On Apr 18, 2016, at 9:35 AM, Ville Kallioniemi <notifications@github.com<mailto:notifications@github.com>> wrote:


@vrv<https://github.com/vrv> Absolutely! I'll start rebasing tonight.

-
You are receiving this because you commented.
Reply to this email directly or view it on GitHub<https://github.com/tensorflow/tensorflow/pull/664#issuecomment-211380353>

ville-k committed:
Fix breaking changes brought in by rebase.

ville-k committed:
* remove redundant cuda/cudnn version functionality
* minimize unnecessary diffs against master branch to ease review/merge

ville-k commented:
@vrv I rebased my branch and removed some redundant functionality. The latest version of apple clang does [not appear to work](https://devtalk.nvidia.com/default/topic/879431/nvcc-amp-clang-7-no-typo-here-/?offset=20) with Cuda 7.5 and you'll have to downgrade to version 7.0.2. I'm using these configure options and bazel incantations to successfully build on OSX:

1. `./configure` with these options
    Host compiler: /usr/bin/clang
    Cuda SDK version: 7.5
    Cuda SDK location: /Developer/NVIDIA/CUDA-7.5
    CuDnn version: 6.5
    CuDnn location: /Developer/NVIDIA/CUDA-6.5

2. `bazel build -c opt --config=cuda //tensorflow/tools/pip_package:build_pip_package `
3. `./bazel-bin/tensorflow/tools/pip_package/build_pip_package /tmp/tensorflow_pkg`
4. `pip install /tmp/tensorflow_pkg/tensorflow-0.8.0rc0-py3-none-any.whl`

Mistobaan commented:
Nice! I tested on my machine with cudnn v5.0 and it works. [Updated the tutorial to help installing the previous version of Xcode 7.2 to make it compile](https://medium.com/@fabmilo/how-to-compile-tensorflow-with-cuda-support-on-osx-fd27108e27e1#.lqbb05men)

royalstream commented:
Regarding the tutorial, this may be worth mentioning: if you don't use Xcode and you only have the command line tools the downgrade is quite simple (delete the **/Library/Developer/CmomandLineTools** folder and install the previous version from Apple's [developer downloads](https://developer.apple.com/downloads))

vrv commented on configure:
```diff
@@ -114,7 +128,12 @@ done
   # Configure the Cudnn version to use.
   if [ -z "$TF_CUDNN_VERSION" ]; then
-    read -p "Please specify the Cudnn version you want to use. [Leave empty to use system default]: " TF_CUDNN_VERSION
+    default_cudnn_version='6.5'
```
the default cudnn version is now 4.0.7 I think

ville-k replied:
Good catch - I had renamed the lib during development. The naming of the lib is a bit inconsistent between platforms. I'll make the default to be a platform dependent: 4.0.7 on Linux and 4 on OS X.

vrv commented on tensorflow/stream_executor/cuda/cuda_diagnostics.cc:
```diff
@@ -61,15 +71,16 @@ string DriverVersionStatusToString(port::StatusOr<DriverVersion> version) {
 
   int major;
   int minor;
+  int patch(0);
```
i think we prefer the style: int patch = 0;

ville-k replied:
Changing.

vrv commented on third_party/gpus/cuda/platform.bzl:
```diff
@@ -0,0 +1,39 @@
+CUDA_VERSION = "7.0"
```
This has changed from being default of "" to "7.0", which I think is misleading.  Is there a reason why you filled in a default again?

ville-k replied:
I had accidentally checked in this file after running configure. Will change CUDA_VERSION, CUDNN_VERSION and PLATFORM to be empty to make this fail loudly if a bug creeps into configure.

vrv commented:
Doh, sorry for not getting to this -- looks like there are a few more conflicts.  I left some comments but this generally looks great -- nice refactoring!

cc @zheng-xq

ville-k committed:
Address code review comments:

* ensure configure generated platform.bzl values default to empty
* fix cudnn default version

ville-k committed:
Fix syntax error introduced during rebase.

ville-k commented:
@vrv Rebased and addressed code review comments.

vrv commented on configure:
```diff
@@ -78,10 +78,21 @@ done
   if [ -z "$TF_CUDA_VERSION" ]; then
-    read -p "Please specify the Cuda SDK version you want to use, e.g. 7.0. [Leave empty to use system default]: " TF_CUDA_VERSION
+    if [ "$OSNAME" == "Linux" ]; then
+      default_cuda_version='7.0'
```
this is backwards, right?

ville-k replied:
I'm not sure what you're referring to,  can you clarify?

vrv replied:
Isn't the default version on Linux 7.5 and the default version on OS X 7.0 ?  You have it the other way here.

ville-k replied:
Oh, sounds like Linux has caught up;) When I created the pull request,  master was using 7.0 with Linux, but OS X clang would only work with 7.5.  I'll make both default to 7.5.

vrv commented on tensorflow/core/platform/load_library.cc:
```diff
@@ -49,6 +49,24 @@ Status GetSymbolFromLibrary(void* handle, const char* symbol_name,
 }
 
+string FormatLibraryFileName(const string& name, const string& version) {
+  string file_name;
```
nit: maybe just filename instead of file_name?

ville-k replied:
Changing.

vrv commented on tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:
```diff
@@ -194,7 +197,15 @@ bool CUDAExecutor::FindOnDiskForComputeCapability(
   char exe_path[PATH_MAX] = {0};
-  CHECK_ERR(readlink("/proc/self/exe", exe_path, sizeof(exe_path) - 1));
+#if defined(__APPLE__)
+    uint32_t buffer_size(0U);
```
same buffer_size = 0U here?

ville-k replied:
Agreed.

vrv commented on third_party/gpus/cuda/platform.bzl:
```diff
@@ -0,0 +1,39 @@
+CUDA_VERSION = ""
```
(can you use two space indents in this file for consistency with the other .bzl files? sorry :(

ville-k replied:
NP, much prefer consistency.

vrv commented on tensorflow/stream_executor/cuda/cuda_diagnostics.cc:
```diff
@@ -149,9 +188,12 @@ void Diagnostician::LogDiagnosticInformation() {
   LOG(INFO) << "kernel reported version is: "
             << DriverVersionStatusToString(kernel_version);
+
+#if !defined(__APPLE__)
```
This is probably worth adding a comment on why this only applies for non-apple.

ville-k replied:
Agreed.

vrv commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -314,10 +314,12 @@ port::Status CudnnSupport::Init() {
     } else {
       const auto& version = result.ValueOrDie();
       LOG(INFO) << "running driver version: " << DriverVersionToString(version);
+#if !defined(__APPLE__)
```
same here -- is there a similar check for the OS X side of things?

ville-k replied:
Agreed.  I haven't been able to find a public API for getting the kernel driver version that is reported by OSX CUDA preferences panel :(

ville-k committed:
Address code review comments - conform to style and add comments.

vrv commented:
Okay, this generally looks good!  Let's weed out any test failures: test this please, @tensorflow-jenkins

vrv commented on configure:
```diff
@@ -127,25 +146,35 @@ while true; do
+    CUDA_DNN_LIB_PATH="lib/libcudnn.${TF_CUDNN_VERSION}.dylib"
+    CUDA_DNN_LIB_ALT_PATH="libcudnn.${TF_CUDNN_VERSION}.dylib"
   fi
-  if [ -e "$CUDNN_INSTALL_PATH/libcudnn.so${TF_CUDNN_EXT}" -o -e "$CUDNN_INSTALL_PATH/lib64/libcudnn.so${TF_CUDNN_EXT}" ]; then
```
So the error here is: 

ERROR: /workspace/third_party/gpus/cuda/BUILD:94:12: in srcs attribute of cc_library rule //third_party/gpus/cuda:cudnn: file '//third_party/gpus/cuda:lib64/libcudnn.so.' is misplaced here (expected .cc, .cpp, .cxx, .c++, .C, .c, .h, .hh, .hpp, .hxx, .inc, .S, .s, .asm, .a, .pic.a, .lo, .pic.lo, .so, .dylib, .o or .pic.o).

Prior to your change, if the user did not set a version, it would default to libcudnn.so, which is a symlink to whatever the latest installed version is (and this is what our tests use).  Now it's being set to libcudnn.so.${TF_CUDNN_VERSION} and that variable is empty.

Is it possible to switch back to the old mode of doing things, where the default is the empty version and the user can specify a specific version (OS X users will manually type in 4 and 7.5 for the versions).  That will also make sure that existing users who depend on the current behavior aren't broken by this change.

ville-k replied:
I pushed a change that makes the *_library_path functions in platform.bzl handle an empty version string.  I'll take a look tomorrow to see if I can make the configure script work the way it did before.

ville-k replied:
@vrv My latest commit restores the original behavior of defaulting to the symlinked cuda libraries when version is left empty in configure.

markb729 commented:
Good work on the update.

There may still be an issue with cudnn v5 when packaged/installed with pip.

For both cudnn v4 and v5, the compile completes successfully. Invoking an example trainer,
bazel-bin/tensorflow/cc/tutorials_example_trainer, runs successfully with BOTH libraries:
>>>I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcublas.7.5.dylib locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcudnn.5.dylib locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcufft.7.5.dylib locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcuda.dylib locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcurand.7.5.dylib locally
I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:883] OS X does not support NUMA - returning NUMA node zero
I tensorflow/core/common_runtime/gpu/gpu_init.cc:102] Found device 0 with properties:
name: GeForce GTX 980 Ti
major: 5 minor: 2 memoryClockRate (GHz) 1.076
...
000001/000009 lambda = 2.000000 x = [0.894427 -0.447214] y = [1.788854 -0.894427]

However, when packaged and installed with pip, a compile with cudnn 4 works as expected but cudnn 5 will fail on load of the cudnn.5.dylib library:

>>> import tensorflow as tf
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcublas.7.5.dylib locally
Segmentation fault: 11

This is the point at which libcudnn.5.dylib loads. Since cudnn 5 works with the example trainer, the packaging must be breaking something, perhaps a misplaced non-versioned symlink? Odd that cudnn 4 works though. Perhaps a misplaced non-versioned symlink?

vrv commented:
@markb729: right now we don't package one binary that works with both cudnn4 and 5, because the APIs are different.

At some point I think it would be conceivable to implement the stream executor in such a way that different cudnn versions are different implementations of the same interface, and we dispatch exactly once during initialization to the right one.

ville-k committed:
Handle empty library versions.

ville-k committed:
Default to symlinked versions of CUDA libraries when no version is specified.

vrv commented on third_party/gpus/cuda/platform.bzl:
```diff
@@ -0,0 +1,39 @@
+  if PLATFORM == "Darwin":
+    return "lib/lib{}.{}.dylib".format(name, version)
+  else:
+    return "lib64/lib{}.so.{}".format(name, version)
```
I think the problem is that you are assuming the version is always set -- but we allow people to use the unversioned library (libfoo.so, not libfoo.so.version).

I'll kick off a test just to make sure, but I still think this may not work.

ville-k replied:
Looks like you were viewing an outdated diff - this was fixed in a previous commit: https://github.com/tensorflow/tensorflow/pull/664/commits/0933043bf4691c72ae3af95d5df3b7e197a3516f

vrv replied:
Sorry, I still am terrible at github.

vrv commented:
@tensorflow-jenkins test this please

vrv commented on third_party/gpus/cuda/cuda_config.sh:
```diff
@@ -99,59 +111,82 @@ function CheckAndLinkToSrcTree {
-elif test -e ${CUDNN_INSTALL_PATH}/include/cudnn.h; then
-  CUDNN_HEADER_PATH=${CUDNN_INSTALL_PATH}/include
-elif test -e /usr/include/cudnn.h; then
-  CUDNN_HEADER_PATH=/usr/include
```
Is this possibly the cause of the test failure?
ERROR: cannot find cudnn.h under: /usr/lib/x86_64-linux-gnu

ville-k replied:
I bet! looks like I rebased that final elif out of existence:)

ville-k replied:
@vrv should be fixed now .

ville-k committed:
Add cudnn.h header search path accidentally lost during rebase.

vrv commented:
One more try!   test this please

vrv commented:
Woohoo!!  thank you so much for this contribution.  We'll try our best to keep it working, though without OS X / GPU test machines, we can't promise too much.

ville-k commented:
@vrv Awesome! I really appreciate your thoughtful feedback and help in figuring out the build issues!

martinwicke commented:
Finally! Thanks for all the hard work! I will get an external GPU so we can test this and make sure it continues to work.

Mistobaan commented:
Nice ! 👍

chrhansen commented:
Hi guys, I'm new here. I was wondering if there shouldn't be a "MacOS **G**PU Tests" in the lists of tests being run at http://ci.tensorflow.org to make sure the new CUDA/GPU functionality keeps working?

martinwicke commented:
Yes. We're installing hardware for that over the weekend. A test should
come some time next week.
On Fri, May 13, 2016 at 17:09 Christian Hansen <notifications@github.com>
wrote:

> Hi guys, I'm new here. I was wondering if there shouldn't be a "MacOS *G*PU
> Tests" in the lists of tests being run at http://ci.tensorflow.org to
> make sure the new CUDA/GPU functionality keeps working?
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/664#issuecomment-219187149>
>

jstaker7 commented:
I'm trying with CUDA 8.0, CuDNN 5.0, and clang-703.0.31 (CUDA test projects seem to build just fine).

I get the follow error:

INFO: Found 1 target...
INFO: From Executing genrule //third_party/gpus/cuda:cuda_config_check [for host]:
/bin/bash: greadlink: command not found
ERROR: /Projects/tensorflow/third_party/gpus/cuda/BUILD:204:1: declared output 'third_party/gpus/cuda/cuda.config' was not created by genrule. This is probably because the genrule actually didn't create this output, or because the output was a directory and the genrule was run remotely (note that only the contents of declared file outputs are copied from genrules run remotely).
ERROR: /Projects/tensorflow/third_party/gpus/cuda/BUILD:204:1: not all outputs were created.
Target //tensorflow/tools/pip_package:build_pip_package failed to build
INFO: Elapsed time: 120.460s, Critical Path: 0.23s

Any ideas?

yjmade commented:
@jstaker7  you need to ```brew install coreutils```

martin-gorner commented:
Since Apple does not sell any computers with Nvidia GPUs,  could you tell us what hardware you are using this with ? Is is some sort of Thunderbolt-attached GPU enclosure with an Nvidia card in it ?

jstaker7 commented:
Once-upon-a-time Apple did provide computers with NVIDIA chips. I have a 2012 MacBook Pro with a 650M.

esd100 commented:
Yes. I have the same the same MacBook pro with 650m card.

royalstream commented:
I still hope Apple is going to offer NVIDIA as an option in some future Macs, if not this year maybe next year.
And of course there's also [iBuildMacs.com](https://ibuildmacs.com/) and [create.pro](https://create.pro/)

yaroslavvb commented:
@martin-gorner Apple sells refurbished laptops with NVidia, ie, http://www.apple.com/shop/product/FE294LL/A/refurbished-154-inch-macbook-pro-23ghz-quad-core-intel-i7-with-retina-display

martin-gorner commented:
Has anyone tried this with GPU in a Thunderbolt enclosure ? This video seems to imply that CUDA works well in that situation: https://youtu.be/Bsf9lHM8qLk

martinwicke commented:
Our (new) Mac GPU tests run in just such a setup: Mac Pro + Quadro M4000 in
a Bizon2.
On Thu, Jul 28, 2016 at 01:56 martin-gorner <notifications@github.com>
wrote:

> Has anyone tried this with GPU in a Thunderbolt enclosure ? This video
> seems to imply that CUDA works well in that situation:
> https://youtu.be/Bsf9lHM8qLk
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/664#issuecomment-235838384>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAjO_Y_Kae-uJER4i9bXzTIJg8y5tqBdks5qaG6ggaJpZM4G9X8_>
> .
>

dandy27 commented:
> Once-upon-a-time Apple did provide computers with NVIDIA chips. I have a 2012 MacBook Pro with a 650M.

i have the same mac pro. do you have it using gpu?i had been trying for 3 days trying to find good tutorials to make it work but nothing.
